classes/local/store/object_file_system.php

```diff
@@ -194,7 +194,15 @@ public function is_file_readable_externally_by_hash($contenthash) {
         $path = $this->get_external_path_from_hash($contenthash, false);
 
         // Note - it is not possible to perform a content recovery safely from a hash alone.
-        return is_readable($path);
+        if (is_readable($path)) {
+            // File is readable externally, so everything is good here.
+            return true;
+        } else {
+            // Otherwise make a deeper check.
+            // This is needed to figure out if the file is simply missing externally or
+            // there is an issue with external storage.
+            return $this->externalclient->is_file_readable_by_hash($contenthash);
```

Comment:
in the base implementation it simply does the exact same is_readable a second time which seems wasteful.

Wouldn't it be better to move all of this logic purely into the s3 store?

Reply:
Perhaps simply return false in the base implementation?